### PR TITLE
03620 token key management functions

### LINKED ISSUE #3620: HIP 514: Add token management functions for key functions
### Problem

The hts precompile flow is missing support for the key management

### Solution

Add support for the key management

- `getTokenKey(address, uint)`
- ~`updateTokenKeys(address, uint, KeyValue)`~ `updateTokenKeys(address,TokenKey)`

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by nikolovyanko**
This PR exports the updateKeys functionality as a separate component than can be called independently thru the HTSPrecomileContract, it also introduces a precompile implementation for the view function "getTokenKey(address, unit)".

**Detailed Description**:

- A new [AbstractTokenUpdatePrecompile](https://github.com/hashgraph/hedera-services/blob/da7b5a38f6faabdf6542015977473b65b3129019/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractTokenUpdatePrecompile.java) class added, in order to wrap the common run logic between:
  - [TokenUpdatePrecompile](https://github.com/hashgraph/hedera-services/blob/da7b5a38f6faabdf6542015977473b65b3129019/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/TokenUpdatePrecompile.java)
  - The new [TokenUpdateKeysPrecompile](https://github.com/hashgraph/hedera-services/blob/da7b5a38f6faabdf6542015977473b65b3129019/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/TokenUpdateKeysPrecompile.java)
  - And [UpdateTokenExpiryInfoPrecompile](https://github.com/hashgraph/hedera-services/blob/8fd1963aa592c4f2e54bb6735e6421719a3feb94/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/UpdateTokenExpiryInfoPrecompile.java) class which is introduced in PR [3756](https://github.com/hashgraph/hedera-services/pull/3756)
- A new [GetTokenKeyPrecompile](https://github.com/hashgraph/hedera-services/blob/da7b5a38f6faabdf6542015977473b65b3129019/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/GetTokenKeyPrecompile.java) class implementing the getTokenKey function.
- Adding `TokenAndTypeCheckSuite`, `TokenExpiryInfoSuite` and `TokenUpdatePrecompileSuite` in `ParallelIntegrationTests`, so that they are enabled in CI/CD

**Related issue(s)**:

Fixes #3620 

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, E2Es)

### CODE REVIEW

**Review by shemnon — APPROVED** [review]

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/AbiConstants.java [hunk 1]
```diff
@@ -112,6 +112,10 @@ private AbiConstants() {
     public static final int ABI_ID_UNFREEZE = 0x52f91387;
     // updateTokenInfo(address token, HederaToken tokenInfo)
     public static final int ABI_ID_UPDATE_TOKEN_INFO = 0x2cccc36f;
+    // updateTokenKeys(address token , TokenKey [])
```
Comment (Nana-EC) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/AbiConstants.java:115]: nit:
```suggestion
    // updateTokenKeys(address token, TokenKey [])
```

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/GetTokenKeyPrecompile.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        final var getTokenKeyWrapper = decoder.decodeGetTokenKey(input);
+        tokenId = getTokenKeyWrapper.tokenID();
+        keyType = getTokenKeyWrapper.tokenKeyType();
+        return super.body(input, aliasResolver);
+    }
+
+    @Override
+    public Bytes getSuccessResultFor(ExpirableTxnRecord.Builder childRecord) {
+        validateTrue(ledgers.tokens().exists(tokenId), ResponseCodeEnum.INVALID_TOKEN_ID);
+        JKey key = (JKey) ledgers.tokens().get(tokenId, keyType);
```
Comment (Nana-EC) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/GetTokenKeyPrecompile.java:59]: Add a `Objects.requireNonNull` check for `keyType`

**Review by Nana-EC — COMMENTED** [review]
LG, minor improvement suggestion

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/EncodingFacade.java [hunk 3]
```diff
@@ -56,38 +66,41 @@
...
-//    private static final TupleType hapiIsApprovedForAllType =
-//            TupleType.parse(INT_BOOL_PAIR_RETURN_TYPE);
-//    private static final TupleType getTokenDefaultFreezeStatusType =
-//            TupleType.parse(INT_BOOL_PAIR_RETURN_TYPE);
-//    private static final TupleType getTokenDefaultKycStatusType =
-//            TupleType.parse(INT_BOOL_PAIR_RETURN_TYPE);
-//    private static final TupleType isTokenFrozenType = TupleType.parse(INT_BOOL_PAIR_RETURN_TYPE);
-//    private static final TupleType isTokenType = TupleType.parse(INT_BOOL_PAIR_RETURN_TYPE);
-//    private static final TupleType getTokenType = TupleType.parse("(int32,int32)");
+    //    private static final String STRING_RETURN_TYPE = "(string)";
```
Comment (shemnon) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/EncodingFacade.java:69]: Is there a need for this commented code?  Should it be removed?

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/utils/DescriptorUtils.java [hunk 4]
```diff
@@ -47,7 +48,9 @@ public static boolean isViewFunction(final Bytes input) {
                     ABI_ID_IS_FROZEN,
                     ABI_ID_IS_KYC,
                     ABI_ID_GET_TOKEN_CUSTOM_FEES,
-                    ABI_ID_GET_TOKEN_KEY -> true;
+                    ABI_ID_GET_TOKEN_KEY,
```
Comment (shemnon) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/utils/DescriptorUtils.java:51]: default should be last

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/EncodingFacade.java (hunk 3 above)
Comment (Nana-EC) [reply hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/EncodingFacade.java:69]: Please remove if no longer needed

**Review by Nana-EC — APPROVED** [review]

**Review by shemnon — APPROVED** [review]

**Comment by Nana-EC** [comment]
@IvanKavaldzhiev please address the final comments so we can get this in.
@Kalina-Todorova

**Comment by IvanKavaldzhiev** [comment]
> @nikolovyanko please address the final comments so we can get this in. @Kalina-Todorova

Hey @Nana-EC , I think everything is resolved. I do not see any pending comment or am I missing something?

**Comment by Nana-EC** [comment]
> > @nikolovyanko please address the final comments so we can get this in. @Kalina-Todorova
> 
> Hey @Nana-EC , I think everything is resolved. I do not see any pending comment or am I missing something?

Indeed, seems everything is resolved now so you should be good to merge

**Review by Nana-EC — APPROVED** [review]
LG

**Pull request merged**